Firmware/ConfigurationStore.cpp

```diff
@@ -268,6 +276,11 @@ bool Config_RetrieveSettings()
                 memcpy_P(&cs.max_acceleration_units_per_sq_second_silent[i],&default_conf.max_acceleration_units_per_sq_second_silent[i],sizeof(cs.max_acceleration_units_per_sq_second_silent[i]));
             }
         }
+        // Initialize arc interpolation settings if they are not already (Not sure about this bit, please review)
+        if (0xff == cs.mm_per_arc_segment) cs.mm_per_arc_segment = DEFAULT_MM_PER_ARC_SEGMENT;
```

Comment:
The default for float values is not handled correctly. Look above at how it was done for `cs.max_feedrate_silent`

Reply:
I was pretty sure I botched that bit.  Are the int settings handled properly?

I will examine ```cs.max_feedrate_silent``` and make the appropriate changes.

Reply:
```
for(uint8_t j = 0; j < sizeof(float); ++j)
            {
                if(0xff != reinterpret_cast<uint8_t*>(&(cs.max_feedrate_silent[i]))[j]) initialized = true;
            }
            if (!initialized) memcpy_P(&cs.max_feedrate_silent[i],&default_conf.max_feedrate_silent[i], sizeof(cs.max_feedrate_silent[i]));
```

It is done correctly for this float.

Reply:
As for the int values, it would be better if you would cast the cs.int value to uint16_t and comparing to 0xFFFF.
So it would be something like this:
`if (((uint16_t)cs.min_arc_segments) == 0xFFFF) cs.min_arc_segments = DEFAULT_MIN_ARC_SEGMENTS;`
I'm not entirely sure that the code I just wrote works, but the idea is that you compare to the correct data type. What you are doing is comparing a int16_t value to 0xFF. It should not work correctly.
If cs.min_arc_segments is uninitialized, then it's value would be 0xFFFF or -1. This depends on whether you compare it as a signed value or as an unsigned value.

I'm not sure if what I just said makes much sense. If you have questions, please tell me ;)

Reply:
One more thing. Both arc_segments_per_sec and min_arc_segments don't really need to be int. They can be unsigned int or uint16_t and nothing will change. It might be easier for you with initialization if they were unsigned

Reply:
I'm still a bit confused about the floats initialization, but it looks like the code should loop through each byte of the float and see if some part of it is not equal to 0xff.  I'm assuming this needs to be done because we don't know the size of a float until the firmware is compiled?   

Anyway, things were looking a bit ugly, so I added these two functions definitions to ConfigurationStore.h:

```
bool is_setting_initialized(float setting_value);
bool is_setting_initialized(uint16_t setting_value);
```

with the following implementations.
```
// Returns true if the setting_value has been initialized, false if not.
bool is_setting_initialized(float setting_value)
{
    // Look at each byte of the arc float settings and see if any byte is initialized
    for (uint8_t j = 0; j < sizeof(float); ++j)
    {
        // If any part of the current byte is not == 0xff, the setting_value has been
        // initialized
        if (0xff != reinterpret_cast<uint8_t*>(&setting_value)[j]) return true;
    }
    // Every byte of the setting_value == 0xff, not initialized
    return false;
}

// Returns true if the setting_value has been initialized, false if not.
bool is_setting_initialized(uint16_t setting_value)
{
    return (0xFFFF != setting_value);
}
```
and I have the following in Config_RetrieveSettings():

```
        // Initialize arc interpolation settings if they are not already
        if (!is_setting_initialized(cs.mm_per_arc_segment)) cs.mm_per_arc_segment = DEFAULT_MM_PER_ARC_SEGMENT;
        if (!is_setting_initialized(cs.min_mm_per_arc_segment)) cs.min_mm_per_arc_segment = DEFAULT_MIN_MM_PER_ARC_SEGMENT;
        if (!is_setting_initialized(cs.min_arc_segments)) cs.min_arc_segments = DEFAULT_MIN_ARC_SEGMENTS;
        if (!is_setting_initialized(cs.arc_segments_per_sec)) cs.arc_segments_per_sec = DEFAULT_ARC_SEGMENTS_PER_SEC;
```

Any obvious problems with that?

Reply:
I don’t see any obvious flaw with the code above.
We cycle through all 4 bytes that make the float value and check if they are different than 0xFF. If all are 0xFF, then it is uninitialized.

The reason that we can’t just compare the float to some value is because, unlike an integer, a float that is all 0xFFFFFFFF is not a valid number (NaN). You can’t just compare it to 0xFFFFFFFF as that is an uint32 integer and has a different float representation.

Reply:
Fantastic.  I just tried downgrading to the current release firmware, doing an M502 + M500 to clear settings, then installed my fork, but my old M214 settings were retained.  I guess the old values are just sort of stuck in the EEPROM.  I'm not sure how to test the initialization code now that I have values set in the EEPROM.  Thoughts?

It may be easier if you try it out tomorrow when you give this code a go.  You can just do an M503 and see what values are shown for the M214 command.  If it doesn't look like this:

```
M214 P1.00 S0.50 R20 F0
```

that means my init code is flawed.